### ISSUE TITLE
[analyzer] False positive core.NullPointerArithm on strrchr with identical constant arguments

### ISSUE THREAD

**Issue by earnol** [llvm/llvm-project#209905]
Basically title says all.

 Reproducer strrchr_const_false_positive.c:

```c
 #include <string.h>
 #define FILE_BASENAME (strrchr(__FILE__, '/') ? strrchr(__FILE__, '/') + 1 : __FILE__)
  const char *get_basename(void) {
      return strrchr("/foo/bar.c", '/') ? strrchr("/foo/bar.c", '/') + 1 : "/foo/bar.c";
  }

  void sink(const char *s);
  void test_file_basename(void) {
      sink(FILE_BASENAME);
  }
```

**Reproduction**:

```
clang --analyze -Xanalyzer -analyzer-checker=core.NullPointerArithm strrchr_const_false_positive.c
```

Godbolt [example](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,selection:(endColumn:3,endLineNumber:14,positionColumn:3,positionLineNumber:14,selectionStartColumn:3,selectionStartLineNumber:14,startColumn:3,startLineNumber:14),source:'+%23include+%3Cstring.h%3E%0A%0A+%23define+FILE_BASENAME+(strrchr(__FILE__,+!'/!')+%3F+strrchr(__FILE__,+!'/!')+%2B+1+:+__FILE__)%0A%0A+const+char+*get_basename(void)%0A+%7B%0A+++return+strrchr(%22/foo/bar.c%22,+!'/!')+%3F+strrchr(%22/foo/bar.c%22,+!'/!')+%2B+1+:+%22/foo/bar.c%22%3B%0A+%7D%0A%0A+void+sink(const+char+*s)%3B%0A+void+test_file_basename(void)%0A+%7B%0A+++sink(FILE_BASENAME)%3B%0A+%7D'),l:'5',n:'0',o:'C%2B%2B+source+%231',t:'0')),k:54.95194802148235,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:clang_trunk,filters:(b:'1',binary:'1',binaryObject:'0',commentOnly:'1',debugCalls:'1',demangle:'1',directives:'1',execute:'0',intel:'0',libraryCode:'1',trim:'1',verboseDemangling:'0'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,libs:!(),options:'--analyze+-Xanalyzer+-analyzer-checker%3Dcore.NullPointerArithm',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+x86-64+clang+(trunk)+(Editor+%231)',t:'0'),(h:output,i:(compilerName:'x86-64+clang+3.4+(assertions)',editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+x86-64+clang+(trunk)+(Compiler+%231)',t:'0')),header:(),k:45.04805197851764,l:'4',m:100,n:'0',o:'',s:1,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4).

**Observed**:

```
warning: Addition of a null pointer and a nonzero integer value results in undefined behavior [core.NullPointerArithm]
```

strrchr_const_false_positive.c: No warning. strrchr(X, C) called twice with the same constant string literal X and same character C must return the same value. If the ternary condition proves the first call is non-null, the second call in the true-branch cannot be null.

Regression from: #207267 ("[analyzer] Model strchr/strrchr/memchr/strstr/strpbrk/strchrnul", commit a34cb573eae6)

**Comment by steakhal** [comment]
Right. Well, I took a shortcut in the implementation that I didn't care about the case when both operands are constant literals - because I figure that should be pretty rare.
Consequently, I decided to split the path and simulate the failure branch regardless of the arguments.
This is what causes your FP.

Speaking of that implementation, it is critical to split the path, because otherwise the model is unsound - so I think the solution is to handle the (probably rare) case of two constant arguments.

**Comment by earnol** [comment]
@steakhal I have made a draft fix for this issue as your commit also cause CSA crash in our code :( To save efforts i'll be not fully reducing crash as it happens only in CTU mode, however i'm going to fix cash as well.

**Comment by steakhal** [comment]
@earnol I have a downstream patch already. Please givve me time so that I could polish it during the weekend.

I'd like to learn more about the crash btw. At least the stack trace.

**Comment by earnol** [comment]
The crash stack trace is:
```
 #0 0x000000000479cf4d llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/project/llvm-project/bin/clang-23+0x479cf4d)
 #1 0x00000000047998c8 llvm::sys::RunSignalHandlers() (/project/llvm-project/bin/clang-23+0x47998c8)
 #2 0x000000000479a39d llvm::sys::CleanupOnSignal(unsigned long) (/project/llvm-project/bin/clang-23+0x479a39d)
 #3 0x00000000046c7898 CrashRecoverySignalHandler(int) CrashRecoveryContext.cpp:0:0
 #4 0x00007f78dc653990 __restore_rt (/lib64/libpthread.so.0+0x12990)
 #5 0x0000000006e48240 (anonymous namespace)::SimpleSValBuilder::simplifySValOnce(llvm::IntrusiveRefCntPtr<clang::ento::ProgramState const>, clang::ento::SVal) (.isra.0) SimpleSValBuilder.cpp:0:0
 #6 0x0000000006e483bc (anonymous namespace)::SimpleSValBuilder::simplifySVal(llvm::IntrusiveRefCntPtr<clang::ento::ProgramState const>, clang::ento::SVal) SimpleSValBuilder.cpp:0:0
 #7 0x0000000006e4b3e0 (anonymous namespace)::SimpleSValBuilder::evalBinOpLN(llvm::IntrusiveRefCntPtr<clang::ento::ProgramState const>, clang::BinaryOperatorKind, clang::ento::Loc, clang::ento::NonLoc, clang::QualType) SimpleSValBuilder.cpp:0:0
 #8 0x00000000068e9a58 (anonymous namespace)::CStringChecker::evalStrchrCommon(clang::ento::CheckerContext&, clang::ento::CallEvent const&, llvm::StringRef, bool) const CStringChecker.cpp:0:0
 #9 0x00000000068e2758 bool clang::ento::eval::Call::_evalCall<(anonymous namespace)::CStringChecker>(void*, clang::ento::CallEvent const&, clang::ento::CheckerContext&) CStringChecker.cpp:0:0
#10 0x0000000006d27d8b clang::ento::CheckerManager::runCheckersForEvalCall(clang::ento::ExplodedNodeSet&, clang::ento::ExplodedNodeSet const&, clang::ento::CallEvent const&, clang::ento::ExprEngine&, clang::ento::EvalCallOptions const&) (/project/llvm-project/bin/clang-23+0x6d27d8b)
#11 0x0000000006da1f8f clang::ento::ExprEngine::evalCall(clang::ento::ExplodedNodeSet&, clang::ento::ExplodedNode*, clang::ento::CallEvent const&) (/project/llvm-project/bin/clang-23+0x6da1f8f)
#12 0x0000000006da332f clang::ento::ExprEngine::VisitCallExpr(clang::CallExpr const*, clang::ento::ExplodedNode*, clang::ento::ExplodedNodeSet&) (/project/llvm-project/bin/clang-23+0x6da332f)
#13 0x0000000006d7b490 clang::ento::ExprEngine::Visit(clang::Stmt const*, clang::ento::ExplodedNode*, clang::ento::ExplodedNodeSet&) (/project/llvm-project/bin/clang-23+0x6d7b490)
#14 0x0000000006d7ce9b clang::ento::ExprEngine::ProcessStmt(clang::Stmt const*, clang::ento::ExplodedNode*) (/project/llvm-project/bin/clang-23+0x6d7ce9b)
#15 0x0000000006d353a7 clang::ento::CoreEngine::dispatchWorkItem(clang::ento::ExplodedNode*, clang::ProgramPoint, clang::ento::WorkListUnit const&) (/project/llvm-project/bin/clang-23+0x6d353a7)
#16 0x0000000006d35a74 clang::ento::CoreEngine::ExecuteWorkList(clang::StackFrame const*, unsigned int, llvm::IntrusiveRefCntPtr<clang::ento::ProgramState const>) (/project/llvm-project/bin/clang-23+0x6d35a74)
#17 0x000000000689d63e (anonymous namespace)::AnalysisConsumer::HandleCode(clang::Decl*, unsigned int, clang::ento::ExprEngine::InliningModes, llvm::DenseSet<clang::Decl const*, llvm::DenseMapInfo<clang::Decl const*, void>>*) AnalysisConsumer.cpp:0:0
#18 0x000000000689f769 (anonymous namespace)::AnalysisConsumer::HandleTranslationUnit(clang::ASTContext&) AnalysisConsumer.cpp:0:0
#19 0x0000000006f6710c clang::ParseAST(clang::Sema&, bool, bool) (/project/llvm-project/bin/clang-23+0x6f6710c)
#20 0x00000000056f543d clang::FrontendAction::Execute() (/project/llvm-project/bin/clang-23+0x56f543d)
#21 0x000000000567cdc0 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/project/llvm-project/bin/clang-23+0x567cdc0)
#22 0x00000000057d615e clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/project/llvm-project/bin/clang-23+0x57d615e)
#23 0x0000000000d8baa9 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/project/llvm-project/bin/clang-23+0xd8baa9)
#24 0x0000000000d80f57 ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&, llvm::ToolContext const&, llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem>) driver.cpp:0:0
#25 0x0000000000d81131 int llvm::function_ref<int (llvm::SmallVectorImpl<char const*>&)>::callback_fn<clang_main(int, char**, llvm::ToolContext const&)::'lambda'(llvm::SmallVectorImpl<char const*>&)>(long, llvm::SmallVectorImpl<char const*>&) driver.cpp:0:0
#26 0x00000000053e2d29 void llvm::function_ref<void ()>::callback_fn<clang::driver::CC1Command::Execute(llvm::ArrayRef<std::optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const::'lambda'()>(long) Job.cpp:0:0
#27 0x00000000046c7d02 llvm::CrashRecoveryContext::RunSafely(llvm::function_ref<void ()>) (/project/llvm-project/bin/clang-23+0x46c7d02)
#28 0x00000000053e3a8b clang::driver::CC1Command::Execute(llvm::ArrayRef<std::optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const (.part.0) Job.cpp:0:0
#29 0x000000000539855b clang::driver::Compilation::ExecuteCommand(clang::driver::Command const&, clang::driver::Command const*&, bool) const (/project/llvm-project/bin/clang-23+0x539855b)
#30 0x000000000539bf7f clang::driver::Compilation::ExecuteJobs(clang::driver::JobList const&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&, bool) const (/project/llvm-project/bin/clang-23+0x539bf7f)
#31 0x00000000053ab66c clang::driver::Driver::ExecuteCompilation(clang::driver::Compilation&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&) (/project/llvm-project/bin/clang-23+0x53ab66c)
#32 0x0000000000d87002 clang_main(int, char**, llvm::ToolContext const&) (/project/llvm-project/bin/clang-23+0xd87002)
#33 0x0000000000c83dcc main (/project/llvm-project/bin/clang-23+0xc83dcc)
#34 0x00007f78d9fd7865 __libc_start_main (/lib64/libc.so.6+0x3a865)
#35 0x0000000000d8048a _start (/project/llvm-project/bin/clang-23+0xd8048a)
clang-23: error: clang frontend command failed due to signal (use -v to see invocation)
```

It happens because call:
```
 State = State->assumeInBound(SymOffset, RemainingExtentBytes, true);
```
might end up returning `NULL`.
I can upload my fix and you can either contribute to my fix or take ideas from it into yours.
